perf(semantic): remove bounds checks on unresolved references stack

overlookmotel opened the pull request:
`SemanticBuilder` keeps a stack of hashmaps to track unresolved references. This stack only grows, never shrinks. Use that invariant to remove bounds checks.

overlookmotel commented:
* **#4390** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/4390?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
More than anything else, this is an experiment to see what effect on perf this kind of optimization has.

In this case, answer seems to be only a little bit, but enter/exit scope isn't a particularly hot path. Likely we could use this same technique on hotter paths for a larger boost.

Boshen commented on crates/oxc_semantic/src/builder.rs:
```diff
@@ -100,18 +95,103 @@ pub struct SemanticBuilderReturn<'a> {
+// to reduce allocations, so the stack grows to maximum scope depth, but never shrinks.
+// See: <https://github.com/oxc-project/oxc/issues/4169>
+// This stack abstraction uses the invariant that stack only grows to avoid bounds checks.
+struct UnresolvedReferencesStack {
```
Move to a new file? Seems out of place.

overlookmotel replied:
Done. NB this is only used in `SemanticBuilder`, not `Semantic`, so I don't think it should be in `scope.rs`.

overlookmotel committed:
perf(semantic): remove bounds checks on unresolved references stack (#4390)

`SemanticBuilder` keeps a stack of hashmaps to track unresolved references. This stack only grows, never shrinks. Use that invariant to remove bounds checks.
